### PR TITLE
cache, goroutine and unbounded workers management

### PR DESCRIPTION
**Pull request by knakul853**
## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [x] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Summary by CodeRabbit

- New Features
  - Shared, configurable caches for compiled regex and DSL expressions.
  - Template cache methods to store/retrieve templates without retaining raw payload data.

- Performance
  - Bounded worker pools for template and HTTP execution to reduce goroutine churn.
  - Fast-path payload loading and pooled buffers for raw HTTP assembly.
  - Regex matching optimizations including literal-prefix short-circuits.

- Bug Fixes
  - Corrected iteration and edge-case handling in matching and extraction logic.

- Tests
  - Added extensive unit tests for caches, generators, matching, payload loading, and execution paths.

### CODE REVIEW

**Review by dwisiswant0 — CHANGES_REQUESTED** [review]
minor: lint error

```
  Error: pkg/operators/matchers/match.go:340:58: SA1019: dslRepo.ErrParsingArg.Error is deprecated: Use errkit.ErrorX instead. (staticcheck)
  	if stringsutil.ContainsAny(err.Error(), "No parameter", dslRepo.ErrParsingArg.Error()) {
```

**Comment by dwisiswant0** [comment]
@coderabbitai list out all references that call the method/func updated by this PR.

File: pkg/templates/compile.go [hunk 1]
```diff
@@ -87,64 +87,64 @@ func Parse(filePath string, preprocessor Preprocessor, options *protocols.Execut
 			// as not using a cache at all, but may be necessary.
 
 			for i, r := range tplCopy.RequestsDNS {
-				rCopy := *r
```
Comment (Ice3man543) [inline pkg/templates/compile.go]: ig this copy of `r` was intentional, because without copy it would modify the reference potentially leading to unintended consequences for `UpdateOptions` we do later.

Earlier we were using direct `r` pointer but it was changed here: https://github.com/projectdiscovery/nuclei/commit/5b89811b906f086e47fdf5357a5ee367f8f1daef#diff-5ff00a91c11854bc967fe7a6dfcee1f0616ee8f3e06ad9715d62e1a848c1b99dR80-R90

File: pkg/operators/extractors/compile.go [hunk 2]
```diff
@@ -4,12 +4,23 @@ import (
...
 
 	"github.com/Knetic/govaluate"
 	"github.com/itchyny/gojq"
 	"github.com/projectdiscovery/nuclei/v3/pkg/operators/common/dsl"
 )
 
+var (
+	extractorRegexCache        sync.Map // map[string]*regexp.Regexp
+	extractorDslCache          sync.Map // map[string]*govaluate.EvaluableExpression
+	extractorMaxRegexCacheSize = 4096
```
Comment (Ice3man543) [inline pkg/operators/extractors/compile.go]: why are we setting max cache size? we should also look into how much the default nuclei templates occupy the cache so we can choose a better number for max performance rather than arbitrary

File: pkg/operators/extractors/compile.go [hunk 3]
```diff
@@ -4,12 +4,23 @@ import (
...
+	"sync"
+	"sync/atomic"
 
 	"github.com/Knetic/govaluate"
 	"github.com/itchyny/gojq"
 	"github.com/projectdiscovery/nuclei/v3/pkg/operators/common/dsl"
 )
 
+var (
+	extractorRegexCache        sync.Map // map[string]*regexp.Regexp
```
Comment (Ice3man543) [inline pkg/operators/extractors/compile.go]: instead of a sync.Map, we can use a expirable-lru or something with a `TTL` or max-count set so we can automatically do evictions / not have to manage the cache size with atomics etc  - https://github.com/hashicorp/golang-lru

File: pkg/operators/matchers/compile.go [hunk 4]
```diff
@@ -5,12 +5,23 @@ import (
...
+	"sync/atomic"
 
 	"github.com/Knetic/govaluate"
 
 	"github.com/projectdiscovery/nuclei/v3/pkg/operators/common/dsl"
 )
 
+var (
+	regexCache        sync.Map // map[string]*regexp.Regexp
+	dslCache          sync.Map // map[string]*govaluate.EvaluableExpression
```
Comment (Ice3man543) [inline pkg/operators/matchers/compile.go]: same here as previous. can we share these somehow to not duplicate

File: pkg/operators/matchers/match.go [hunk 5]
```diff
@@ -106,10 +109,33 @@ func (matcher *Matcher) MatchRegex(corpus string) (bool, []string) {
 	var matchedRegexes []string
 	// Iterate over all the regexes accepted as valid
 	for i, regex := range matcher.regexCompiled {
-		// Continue if the regex doesn't match
-		if !regex.MatchString(corpus) {
-			// If we are in an AND request and a match failed,
-			// return false as the AND condition fails on any single mismatch.
+		// Literal prefix short-circuit
+		rstr := regex.String()
+		if !strings.Contains(rstr, "(?i") { // covers (?i) and (?i:
```
Comment (Ice3man543) [inline pkg/operators/matchers/match.go:111]: we can add tests for each of these individual logics to ensure everything works as intended / no side effects

File: pkg/protocols/common/generators/load.go [hunk 6]
```diff
@@ -17,6 +17,20 @@ func (generator *PayloadGenerator) loadPayloads(payloads map[string]interface{},
 	for name, payload := range payloads {
 		switch pt := payload.(type) {
 		case string:
+			// Fast path: if no newline, treat as file path
+			if !strings.ContainsRune(pt, '\n') {
+				file, err := generator.options.LoadHelperFile(pt, templatePath, generator.catalog)
```
Comment (Ice3man543) [inline pkg/protocols/common/generators/load.go:22]: can we add test for this as well?

File: pkg/protocols/http/raw/raw_test.go [hunk 7]
```diff
@@ -7,6 +7,24 @@ import (
...
+	before := string(r.UnsafeRawBytes)
+	// call again to ensure pooled buffer reuse won't corrupt previous slice
+	if err := r.TryFillCustomHeaders([]string{"X-Another: 2"}); err != nil {
+		t.Fatalf("unexpected err on second call: %v", err)
+	}
+	after := string(r.UnsafeRawBytes)
+	if before == after {
+		t.Fatalf("expected second modification to change bytes; got identical")
+	}
+}
```
[the remark this answers is not in the document]
Comment (Ice3man543) [reply pkg/protocols/http/raw/raw_test.go]: @knakul853 ^

File: pkg/templates/cache.go [hunk 8]
```diff
@@ -1,18 +1,26 @@
...
 	"github.com/projectdiscovery/utils/conversion"
 	mapsutil "github.com/projectdiscovery/utils/maps"
 )
 
 // Templates is a cache for caching and storing templates for reuse.
 type Cache struct {
 	items *mapsutil.SyncLockMap[string, parsedTemplate]
+
+	// parsedTemplatePool reduces allocation overhead when storing cache entries
+	parsedTemplatePool *sync.Pool
```
Comment (Ice3man543) [inline pkg/templates/cache.go]: do we really need this? Ig storing items in mapsutil cache is enoug right as anyway we aren't caching compiled templates?

File: pkg/operators/matchers/regexcache.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+	"sync"
+	"time"
+)
+
+type regexCacheKey struct {
+	pattern string
+	corpus  string
+}
+
+type regexCacheEntry struct {
```
Comment (Ice3man543) [inline pkg/operators/matchers/regexcache.go]: Do we really need to cache matches and does it actually improve because we are storing matched and not sure if its required (as we also incur storage overhead) 

We can do some benchmarks / synthetic tests to identigy whether it improves on standard nuclei-templates run

File: pkg/operators/extractors/compile.go (hunk 2 above)
Comment (knakul853) [reply pkg/operators/extractors/compile.go]: it was random max limit just did audit here its seems 
around  30% of the limit are reached most of the time 
and 4mb seems more than enough. 
regex 3434/4096 and DSL 593/4096, 
Screenshot 2025-09-09 at 8 33 02 PM

File: pkg/operators/matchers/regexcache.go (hunk 9 above)
Comment (knakul853) [reply pkg/operators/matchers/regexcache.go]: its used in the  pkg/operators/matchers/match.go
and pkg/operators/extractors/extract.go 
Screenshot 2025-09-10 at 12 27 50 AM


here its performance impacts, but yeah it reduce cpu overhead but introduce storage one. parking this to be handle next time as its still huge performance bottleneck.

File: pkg/templates/cache.go (hunk 8 above)
Comment (knakul853) [reply pkg/templates/cache.go]: on benchmark, it has very tiny impacts over 10k keys, mostly the items keys grows, so for now removing the pool. though in future specially in case of multiple instance when `items` allocation increases, we can introduce pool

File: pkg/protocols/common/generators/load.go (hunk 6 above)
Comment (knakul853) [reply pkg/protocols/common/generators/load.go:22]: done

File: pkg/operators/matchers/match.go (hunk 5 above)
Comment (knakul853) [reply pkg/operators/matchers/match.go:111]: done

File: pkg/operators/matchers/compile.go (hunk 4 above)
Comment (knakul853) [reply pkg/operators/matchers/compile.go]: done

File: pkg/operators/extractors/compile.go (hunk 3 above)
Comment (knakul853) [reply pkg/operators/extractors/compile.go]: done

File: pkg/templates/compile.go (hunk 1 above)
Comment (knakul853) [reply pkg/templates/compile.go]: this is very expensive, i can see currently we can use safely but seems HDM has added this for make multi instance caching safe.reverting to HDM changes, but its continue to have cpu performance degrade until we start using those.

File: pkg/operators/cache/cache.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+func initCaches() {
+	initOnce.Do(func() {
+		regexCache = gcache.New[string, *regexp.Regexp](regexCap).LRU().Build()
+		dslCache = gcache.New[string, *govaluate.EvaluableExpression](dslCap).LRU().Build()
+	})
+}
+
+func SetCapacities(regexCapacity, dslCapacity int) {
+	// ensure caches are initialized under initOnce, so later Regex()/DSL() won't re-init
+	initCaches()
```
Comment (dwisiswant0) [inline pkg/operators/cache/cache.go:31]: why isn’t this under a lock? can you confirm this isn’t racy?

File: pkg/core/executors.go [hunk 11]
```diff
@@ -78,6 +81,36 @@ func (e *Engine) executeTemplateWithTargets(ctx context.Context, template *templ
...
+	for i := 0; i < workerCount; i++ {
+		go func() {
+			defer workersWg.Done()
+			for t := range tasks {
+				func() {
+					defer cleanupInFlight(t.index)
+					if t.skip {
+						return
+					}
+
```
Comment (dwisiswant0) [inline pkg/core/executors.go:112]: potential UB &/ leaks. pretty sure we need ctx cancellation checks here before exec.

File: pkg/protocols/http/request.go [hunk 12]
```diff
@@ -240,6 +240,43 @@ func (request *Request) executeParallelHTTP(input *contextargs.Context, dynamicV
...
+	// bounded worker-pool to avoid spawning one goroutine per payload
+	type task struct {
+		req          *generatedRequest
+		updatedInput *contextargs.Context
+	}
+
+	var workersWg sync.WaitGroup
+	currentWorkers := maxWorkers
+	tasks := make(chan task, maxWorkers)
+	spawnWorker := func() {
```
Comment (dwisiswant0) [inline pkg/protocols/http/request.go]: we should be tracking workers + making sure they can be cancelled here instead.

(i know L288–292 handles one case, but L277 spawn slips thru)

```suggestion
	spawnWorker := func(ctx context.Context) {
```

File: pkg/operators/cache/cache.go (hunk 10 above)
Comment (knakul853) [reply pkg/operators/cache/cache.go:31]: here SetCapacities isn’t racy. We initialize caches under sync.Once, and both reads and rebuilds are protected by mutax (readers use RLock, writers use Lock). I validated with a race test that concurrently calls Regex/DSL and SetCapacities; it passes with -race. will push this test, have a look.

Comment (dwisiswant0) [reply pkg/operators/cache/cache.go:31]: > We initialize caches under sync.Once, [...]

aight, didn't notice that.

**Review by dwisiswant0 — APPROVED** [review]
